packages/docprovider/src/tokens.ts

```diff
@@ -22,6 +22,11 @@ export interface IDocumentProvider {
    */
   putInitializedState(): void;
 
+  /**
+   * Put the document source.
+   */
+  putSource(source: string): void;
```

Comment:
Since this adds a new (non-optional) method to the `IDocumentProvider` interface, I think this should be considered as a backward incompatible change.

Would you mind adding that to the "Backwards-incompatible changes" section of the PR template so we can keep track of it? Thanks!

Reply:
Sure, although we could make the method optional, and fall back to the previous behavior if it doesn't exist?
Also, this PR:
- makes it mandatory to be in collaborative mode, but I guess we could support non-collaborative mode too.
- requires a change in the server. I made the change [in jupyverse](https://github.com/jupyter-server/jupyverse/pull/127), but it should also be done in `jupyter-server`.

Reply:
I think that we can remove this method eventually as we will be able to compute the source using y-py. This is just meant as an intermediate step, right?

The HTTP API shouldn't be able to overwrite the current RTC state. Hence, the PUT request shouldn't contain any content when RTC is enabled and simply grab the source from the RTC server and then save it to disk.

To maintain a bit of backwards compatibility we could still make the `content` value in the PUT request optional:

* If there currently exists an RTC session, and `content` is empty. Calculate the source based on the RTC state and save it to disk (for now we will use the stored `source` value that David implemented here).
* If there currently exists an RTC session, and `content` is **not** empty. Return a 409 error code.
* If there is no active RTC session for that room, and `content` is **not** empty. Save the `content` as usual.
* Lastly. if there is no active RTC session, and content is not empty, return an error code. This scenario should never happen as the sending client is clearly an RTC client, but for some reason the RTC session is not available anymore. Maybe a 410 might be appropriate.

Reply:
For the last point, do you mean "if there is no active RTC session, and content is empty"?